contracts/multisig-prover/src/contract.rs

```diff
@@ -1,44 +1,479 @@
+    WasmMsg, WasmQuery,
+};
+use cw_utils::{parse_reply_execute_data, MsgExecuteContractResponse};
+
+use std::{collections::HashMap, str::FromStr};
+
+use axelar_wasm_std::{Participant, Snapshot};
+use connection_router::{msg::Message, types::ChainName};
+use multisig::{msg::Multisig, types::MultisigState};
+use service_registry::state::Worker;
 
 use crate::{
     error::ContractError,
+    events::Event,
     msg::ExecuteMsg,
-    msg::{GetProofResponse, QueryMsg},
+    msg::{GetProofResponse, InstantiateMsg, ProofStatus, QueryMsg},
+    state::{Config, COMMANDS_BATCH, CONFIG, KEY_ID, PROOF_BATCH_MULTISIG, REPLY_BATCH},
+    types::{BatchID, CommandBatch, ProofID},
 };
 
+const START_MULTISIG_REPLY_ID: u64 = 1;
+
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn execute(
-    _deps: DepsMut,
+pub fn instantiate(
+    deps: DepsMut,
     _env: Env,
     _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let admin = deps.api.addr_validate(&msg.admin_address)?;
+    let gateway = deps.api.addr_validate(&msg.gateway_address)?;
+    let multisig = deps.api.addr_validate(&msg.multisig_address)?;
+    let service_registry = deps.api.addr_validate(&msg.service_registry_address)?;
+
+    let config = Config {
+        admin,
+        gateway,
+        multisig,
+        service_registry,
+        destination_chain_id: msg.destination_chain_id,
+        signing_threshold: msg.signing_threshold,
+        service_name: msg.service_name,
+        chain_name: ChainName::from_str(&msg.chain_name).map_err(
+            |e: connection_router::ContractError| ContractError::InvalidInput {
+                reason: e.to_string(),
+            },
+        )?,
+    };
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
     msg: ExecuteMsg,
 ) -> Result<Response, ContractError> {
     match msg {
-        ExecuteMsg::ConstructProof { message_ids } => execute::construct_proof(message_ids),
+        ExecuteMsg::ConstructProof { message_ids } => execute::construct_proof(deps, message_ids),
+        ExecuteMsg::RotateSnapshot { pub_keys, key_id } => {
+            execute::rotate_snapshot(deps, env, info, pub_keys, key_id)
+        }
     }
 }
 
 pub mod execute {
     use super::*;
 
-    pub fn construct_proof(_message_ids: Vec<String>) -> Result<Response, ContractError> {
-        todo!()
+    pub fn construct_proof(
+        deps: DepsMut,
+        message_ids: Vec<String>,
+    ) -> Result<Response, ContractError> {
+        let key_id = KEY_ID.load(deps.storage)?;
+
+        let config = CONFIG.load(deps.storage)?;
+
+        let batch_id = BatchID::new(&message_ids);
+
+        let query = gateway::msg::QueryMsg::GetMessages { message_ids };
+        let messages: Vec<Message> = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
```

Comment:
What does the gateway do in the query if it cannot find some of the given message ids?

Reply:
as it is right now, it will fail